dart/lib/src/exception_cause_extractor.dart

```diff
@@ -0,0 +1,40 @@
+import 'exception_cause.dart';
+import 'sentry_options.dart';
+import 'throwable_mechanism.dart';
+
+abstract class ExceptionCauseExtractor<T> {
+  ExceptionCause? cause(T error);
+  Type get exceptionType => T;
+}
+
+class RecursiveExceptionCauseExtractor {
+  RecursiveExceptionCauseExtractor(this._options);
+
+  final SentryOptions _options;
+
+  List<ExceptionCause> flatten(exception, stackTrace) {
+    final allExceptionCauses = <ExceptionCause>[];
+    final circularityDetector = <dynamic>{};
+
+    var currentException = exception;
+    ExceptionCause? currentExceptionCause =
+        ExceptionCause(exception, stackTrace);
+
+    while (currentException != null &&
+        currentExceptionCause != null &&
+        circularityDetector.add(currentException)) {
+      allExceptionCauses.add(currentExceptionCause);
+
+      final extractionSourceSource = currentException is ThrowableMechanism
+          ? currentException.throwable
+          : currentException;
+
+      final extractor =
+          _options.exceptionCauseExtractor(extractionSourceSource.runtimeType);
+
+      currentExceptionCause = extractor?.cause(extractionSourceSource);
```

Comment:
Better to wrap that with a try-catch, otherwise, we will silently swallow the exception in case the extract has a bug.
Logging the error too.